### PR TITLE
[#11571] Cascade user updates/deletes to deadline maps

### PR DESCRIPTION
**Pull request by jayasting98**
Part of #11571 

**PR Checklist**

**Outline of Solution**

Any user email address update or any user deletion is cascaded down to the deadline maps. This is implemented by finding all the feedback sessions in the course the user belongs to. For each feedback session, the email is updated or deleted accordingly in the respective deadline map if that map contains a selective deadline for that user.

### CODE REVIEW

File: src/test/resources/data/typicalDataBundle.json [hunk 1]
```diff
@@ -816,7 +816,9 @@
       "studentDeadlines": {
         "student4InCourse1@gmail.tmt": "2026-04-28T23:00:00Z"
       },
-      "instructorDeadlines": {}
+      "instructorDeadlines": {
+        "instructor1@course1.tmt": "2027-04-30T23:00:00Z"
+      }
```
Comment (samuelfangjw) [inline src/test/resources/data/typicalDataBundle.json:821]: For the test json files, let's add the corresponding DeadlineExtensions as well.

**Review by samuelfangjw — COMMENTED** [review]
For the FeedbackSessionLogic
1. Let's group the delete methods and update methods together.
2. I don't think we need so many methods for a simple update. The current implementation seems unnecessarily complex for this purpose.

I think something like this will be more suitable. I've tried to write it in your style.

```java
    public void deleteFeedbackSessionsDeadlinesForUser(String courseId, String email, boolean isInstructor) {
        Consumer<Map<String, Instant>> deadlineUpdater = deadlines -> deadlines.remove(email);

        updateFeedbackSessionDeadlinesForUser(courseId, email, isInstructor, deadlineUpdater);
    }

    public void updateFeedbackSessionDeadlinesWithNewEmail(String courseId, String oldEmail, String newEmail, boolean isInstructor) {
        if (oldEmail.equals(newEmail)) {
            return;
        }

        Consumer<Map<String, Instant>> deadlineUpdater = deadlines -> deadlines.put(newEmail, deadlines.remove(oldEmail));

        updateFeedbackSessionDeadlinesForUser(courseId, oldEmail, isInstructor, deadlineUpdater);
    }

    private void updateFeedbackSessionDeadlinesForUser(String courseId, String email,
            boolean isInstructor, Consumer<Map<String, Instant>> deadlineUpdater) {
        List<FeedbackSessionAttributes> feedbackSessions = fsDb.getFeedbackSessionsForCourse(courseId);
        for (var session : feedbackSessions) {
            Map<String, Instant> studentDeadlines = session.getStudentDeadlines();
            Map<String, Instant> instructorDeadlines = session.getInstructorDeadlines();
            if (isInstructor && instructorDeadlines.containsKey(email)) {
                deadlineUpdater.accept(instructorDeadlines);
            } else if (!isInstructor && studentDeadlines.containsKey(email)) {
                deadlineUpdater.accept(studentDeadlines);
            } else {
                // No deadline found in session
                continue;
            }
            try {
                fsDb.updateFeedbackSession(FeedbackSessionAttributes
                        .updateOptionsBuilder(session.getFeedbackSessionName(), courseId)
                        .withStudentDeadlines(studentDeadlines)
                        .withInstructorDeadlines(instructorDeadlines)
                        .build());
            } catch (InvalidParametersException | EntityDoesNotExistException e) {
                // Both Exceptions should not be thrown
                log.severe("Unexpected error", e);
            }
        }
    }
```

**Comment by jayasting98** [comment]
@samuelfangjw I initially did it similarly to how you suggested, also to mirror your implementation for cascading for the deadline extension entity since it seemed pretty good. However, I felt that for your case, the reason why cascading to the entity can be implemented that way is because `isInstructor` is actually a property of the entity. On the other hand, in my case, all it was doing was changing the behaviour of the method, so I thought that meant `isInstructor` would be a [flag argument](https://martinfowler.com/bliki/FlagArgument.html). I decided to split it into different methods instead to avoid that. What are your thoughts on this?

Regardless, I think for right now, I will change `deadlinesUpdater` into a `Consumer` like you suggested because that seems much better.

**Comment by samuelfangjw** [comment]
> @samuelfangjw I initially did it similarly to how you suggested, also to mirror your implementation for cascading for the deadline extension entity since it seemed pretty good. However, I felt that for your case, the reason why cascading to the entity can be implemented that way is because `isInstructor` is actually a property of the entity. On the other hand, in my case, all it was doing was changing the behaviour of the method, so I thought that meant `isInstructor` would be a [flag argument](https://martinfowler.com/bliki/FlagArgument.html). I decided to split it into different methods instead to avoid that. What are your thoughts on this?

If you can do it in a way that doesn't sacrifice readability I don't see any issues. 

My personal thoughts are that we should have a pragmatic approach to refactoring and make sure any "rules" we apply actually makes sense in the context.

File: src/test/java/teammates/logic/core/FeedbackSessionsLogicTest.java [hunk 2]
```diff
@@ -700,6 +704,182 @@ public void testUpdateFeedbackSession_shouldAdjustEmailSendingStatusAccordingly(
                 typicalSession.getFeedbackSessionName(), typicalSession.getCourseId()).isSentPublishedEmail());
     }
 
+    @Test
+    public void testUpdateFeedbackSessionsInstructorDeadlinesWithNewEmail() {
```
Comment (moziliar) [inline src/test/java/teammates/logic/core/FeedbackSessionsLogicTest.java:708]: Maybe the test case name can describe the input and the output, say `updateInstructorEmail_transferDeadlinesToNewEmail`?

File: src/test/java/teammates/logic/core/FeedbackSessionsLogicTest.java [hunk 3]
```diff
@@ -700,6 +704,182 @@ public void testUpdateFeedbackSession_shouldAdjustEmailSendingStatusAccordingly(
...
+                    return counts;
+                }, (curr, next) -> {
+                    curr.putAll(next);
+                    return curr;
+                });
+        assertEquals(oldDeadlineCounts, newDeadlineCounts);
+    }
+
+    @Test
+    public void testUpdateFeedbackSessionsStudentDeadlinesWithNewEmail() {
```
Comment (moziliar) [inline src/test/java/teammates/logic/core/FeedbackSessionsLogicTest.java:755]: ditto

File: src/test/java/teammates/logic/core/FeedbackSessionsLogicTest.java [hunk 4]
```diff
@@ -700,6 +704,182 @@ public void testUpdateFeedbackSession_shouldAdjustEmailSendingStatusAccordingly(
...
+        String courseId = student4InCourse1.getCourse();
+        String oldEmailAddress = student4InCourse1.getEmail();
+        String newEmailAddress = "new@email.tmt";
+
+        Map<Instant, Integer> oldDeadlineCounts = fsLogic.getFeedbackSessionsForCourse(courseId)
+                .stream()
+                .map(FeedbackSessionAttributes::getStudentDeadlines)
+                .filter(studentDeadlines -> studentDeadlines.containsKey(oldEmailAddress))
+                .map(studentDeadlines -> studentDeadlines.get(oldEmailAddress))
+                .reduce(new HashMap<>(), (counts, deadline) -> {
```
Comment (moziliar) [inline src/test/java/teammates/logic/core/FeedbackSessionsLogicTest.java:766]: ditto?

File: src/test/java/teammates/logic/core/FeedbackSessionsLogicTest.java [hunk 5]
```diff
@@ -700,6 +704,182 @@ public void testUpdateFeedbackSession_shouldAdjustEmailSendingStatusAccordingly(
...
+        String emailAddress = instructor1OfCourse1.getEmail();
+
+        // The instructor should have selective deadlines.
+        Set<FeedbackSessionAttributes> oldSessionsWithInstructor1Deadlines = fsLogic
+                .getFeedbackSessionsForCourse(courseId)
+                .stream()
+                .filter(feedbackSessionAttributes -> feedbackSessionAttributes.getInstructorDeadlines()
+                        .containsKey(emailAddress))
+                .collect(Collectors.toSet());
+        assertEquals(2, oldSessionsWithInstructor1Deadlines.size());
```
Comment (moziliar) [inline src/test/java/teammates/logic/core/FeedbackSessionsLogicTest.java:816]: Can consider move to after line 819?

**Review by moziliar — CHANGES_REQUESTED** [review]

File: src/main/java/teammates/logic/core/FeedbackSessionsLogic.java [hunk 6]
```diff
@@ -568,4 +600,48 @@ public boolean isFeedbackSessionForUserTypeToAnswer(FeedbackSessionAttributes se
...
+        updateFeedbackSessionsDeadlinesForUser(courseId, oldEmailAddress, isInstructor,
+                deadlines -> deadlines.put(newEmailAddress, deadlines.remove(oldEmailAddress)));
+    }
+
+    private void deleteFeedbackSessionsDeadlinesForUser(String courseId, String emailAddress, boolean isInstructor) {
+        updateFeedbackSessionsDeadlinesForUser(courseId, emailAddress, isInstructor,
+                deadlines -> deadlines.remove(emailAddress));
+    }
+
+    private void updateFeedbackSessionsDeadlinesForUser(String courseId, String emailAddress, boolean isInstructor,
```
Comment (moziliar) [inline src/main/java/teammates/logic/core/FeedbackSessionsLogic.java:617]: This looks quite readable to me! Kudos!

Comment (jayasting98) [reply src/main/java/teammates/logic/core/FeedbackSessionsLogic.java:617]: Kudos instead to Samuel lol, this is more of his code. I just modified his original suggestion a bit here and there.

File: src/test/java/teammates/logic/core/FeedbackSessionsLogicTest.java (hunk 2 above)
Comment (jayasting98) [reply src/test/java/teammates/logic/core/FeedbackSessionsLogicTest.java:708]: I think the test method name gets a bit too long if I add it to the end:

```
public void testUpdateFeedbackSessionsInstructorDeadlinesWithNewEmail_updateInstructorEmail_transferDeadlinesToNewEmail()
```

Or do you mean replacing it completely? Wouldn't this be unconventional?

```
public void testUpdateInstructorEmail_transferDeadlinesToNewEmail()
```

Or how about using something like `______TS("Update instructor email; transfers deadlines to new email.")`?

Comment (moziliar) [reply src/test/java/teammates/logic/core/FeedbackSessionsLogicTest.java:708]: I think TS will do. Indeed it isn't too conventional to just replace as such.

File: src/test/java/teammates/logic/core/FeedbackSessionsLogicTest.java [hunk 7]
```diff
@@ -700,6 +704,162 @@ public void testUpdateFeedbackSession_shouldAdjustEmailSendingStatusAccordingly(
...
+
+        ______TS("Update email; transfers deadlines to new email.");
+
+        Map<Instant, Long> oldDeadlineCounts = fsLogic.getFeedbackSessionsForCourse(courseId)
+                .stream()
+                .map(FeedbackSessionAttributes::getStudentDeadlines)
+                .filter(studentDeadlines -> studentDeadlines.containsKey(oldEmailAddress))
+                .map(studentDeadlines -> studentDeadlines.get(oldEmailAddress))
+                .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
+        assertEquals(2L, oldDeadlineCounts.values()
```
Comment (moziliar) [inline src/test/java/teammates/logic/core/FeedbackSessionsLogicTest.java:757]: Indentation seems off where the sub stream is mixed up with the parent stream. Also, there's way to make the count remain in int by not using .counting()

File: src/test/java/teammates/logic/core/InstructorsLogicTest.java [hunk 8]
```diff
@@ -317,6 +322,58 @@ private void testVerifyAtLeastOneInstructorIsDisplayed() throws Exception {
...
+        String courseId = instructorToBeUpdated.getCourseId();
+        String oldEmailAddress = instructorToBeUpdated.getEmail();
+        String newEmailAddress = "new@email.tmt";
+
+        Map<Instant, Integer> oldDeadlineCounts = fsLogic.getFeedbackSessionsForCourse(courseId)
+                .stream()
+                .map(FeedbackSessionAttributes::getInstructorDeadlines)
+                .filter(instructorDeadlines -> instructorDeadlines.containsKey(oldEmailAddress))
+                .map(instructorDeadlines -> instructorDeadlines.get(oldEmailAddress))
+                .reduce(new HashMap<>(), (counts, deadline) -> {
```
Comment (moziliar) [inline src/test/java/teammates/logic/core/InstructorsLogicTest.java:337]: A few more places like this are unclear. Can apply the same fix here

**Review by moziliar — CHANGES_REQUESTED** [review]

File: src/test/java/teammates/logic/core/FeedbackSessionsLogicTest.java (hunk 7 above)
Comment (jayasting98) [reply src/test/java/teammates/logic/core/FeedbackSessionsLogicTest.java:757]: I used `summingInt` initially but I thought `counting` was clearer, just that the type had to be changed, but I thought this was not a problem. I can switch back to `summingInt` though.

File: src/test/java/teammates/logic/core/InstructorsLogicTest.java (hunk 8 above)
Comment (jayasting98) [reply src/test/java/teammates/logic/core/InstructorsLogicTest.java:337]: Ah sorry, I forgot I did it in other files too.

File: src/test/java/teammates/logic/core/FeedbackSessionsLogicTest.java (hunk 7 above)
Comment (moziliar) [reply src/test/java/teammates/logic/core/FeedbackSessionsLogicTest.java]: Just some thoughts... It seems ok looking on a second look. Just keep it this way

Comment (jayasting98) [reply src/test/java/teammates/logic/core/FeedbackSessionsLogicTest.java]: Oops, I changed it already. I hope that's fine.

**Review by moziliar — APPROVED** [review]
LGTM! I'm ok with the rest of the changes!

**Review by samuelfangjw — APPROVED** [review]

**Review by halfwhole — APPROVED** [review]
Looks good!

**Pull request merged**